src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,384 @@ public class SqlMutationEngine : IMutationEngine
+            {
+                // The primary keys can have a mapping defined. mutationResultRow contains the mapped column names as keys.
+                // So, the mapped field names are used to look up and fetch the values from mutationResultRow.
+                // TryGetExposedColumnName method populates the the mapped column name (if configured) or the original column name into exposedColumnName.
+                // It returns false if the primary key does not exist.
+                if (sqlMetadataProvider.TryGetExposedColumnName(context.EntityName, primarykey, out string? exposedColumnName))
+                {
+                    findRequestContext.PrimaryKeyValuePairs.Add(exposedColumnName, mutationResultRow.Columns[exposedColumnName]!);
+                }
+                else
+                {
+                    // This code block should never be reached because the information about primary keys gets populated during the startup.
+                    throw new DataApiBuilderException(
+                       message: "Insert/Upsert operation was successful but unexpected error when constructing the response",
+                       statusCode: HttpStatusCode.InternalServerError,
+                       subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedErrorFindingExposedFieldName);
+                }
+            }
+
+            // READ action for the given role can have include and exclude fields configured. Populating UpdateReturnFields
+            // ensures that the select query retrieves only those fields that are allowed for the given role.
+            findRequestContext.UpdateReturnFields(_authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read));
+
+            return findRequestContext;
+        }
+
+        /// <summary>
+        /// Constructs and returns a HTTP 201 Created response.
+        /// The response is constructed using results of the upsert database operation when database policy is not defined for the read permission.
+        /// If database policy is defined, the results of the subsequent select statement is used for constructing the response.
+        /// </summary>
+        /// <param name="resultRow">Reuslt of the upsert database operation</param>
+        /// <param name="jsonDocument">Result of the select database operation</param>
+        /// <param name="primaryKeyRoute">Primary key route to be used in the Location Header</param>
+        /// <param name="isReadPermissionConfiguredForRole">Indicates whether read permissions is configured for the role</param>
+        /// <param name="isDatabasePolicyDefinedForReadAction">Indicates whether database policy is configured for read action</param>
+        /// <param name="operationType">Resultant Operation type - Update, Insert, etc.</param>
+        private CreatedResult ConstructCreatedResultResponse(
```

Comment:
Won't this method make more sense in `SqlResponseHelpers`? Has nothing to do with the mutation engine.

Reply:
All the callers of this (and other similar methods for the similar suggestion) are within the same class, so it appears it should be okay to leave these as is. 

I don't have a strong preference on keeping these methods here vs moving to the `SqlResponseHelpers`. Please let me know of advantages you see in moving them over to the other class [I'm open to moving them, just want to understand your thought process more :) ]

Reply:
Seperation of concerns or SRP( Single Responsibility Principle). Does not matter if this method has only one reference and that too in the SqlMutationEngine. Quoting an example from our repo only, there might be a method like `RequestValidator.ValidateInsertRequestContext` having a single reference, called in `RestService` but the method is doing request validation. Thus, the method is placed in `RequestValidator` only.

Reply:
Cool. Moved the methods to `SqlResponseHelpers` class